### PR TITLE
Updated code adding GEM hits in STA reconstruction

### PR DESCRIPTION
**Pull request by archiesharma**
Closing the old pull request #9806, and creating this new pull request in a new area with all the updates and implementing all the comments from the old pull request #9806 :

1) Hit Pattern changes (replacing GEM rolls with GEM stations) are applied and in the very recent IB release CMSSW_7_6_X_2015-07-04-2300 so as not to conflict with the other changes made in #9959.
2) Untracked parameters in  MuonServiceProxy are changed to tracked.
3) All the commented lines in the code are removed.
4) Tests are done with  runTheMatrix.py -l 11325.0 --useInput all and all of them are ok.

### CODE REVIEW

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by deguio** [comment]
+1

**Comment by archiesharma** [comment]
@cmsbuild  Could you please explain the log mentioned above with error messages, where exactly the code is failing.   Because of very long size of the log files, we are no able to point the reason of the error. We also tested the full branch with "runTheMatrix.py -l 11325.0 --useInput all" and it is passing all the tests.

File: RecoMuon/Navigation/src/MuonNavigationPrinter.cc [hunk 1]
```diff
@@ -52,7 +55,10 @@ MuonNavigationPrinter::MuonNavigationPrinter(const MuonDetLayerGeometry * muonLa
   PRINT("MuonNavigationPrinter")  << "BACKWARD:" << std::endl;
 
   vector<const DetLayer*> backward;
-  if ( enableRPC ) backward = muonLayout->allBackwardLayers();
+  if ( enableCSC & enableGEM & enableRPC ) backward = muonLayout->allBackwardLayers();
+  else if ( enableCSC & enableGEM & !enableRPC ) backward = muonLayout->allCscGemBackwardLayers(); // CSC + GEM
+  else if ( !enableCSC & enableGEM & !enableRPC ) backward = muonLayout->backwardGEMLayers(); //GEM only
+  else if ( enableCSC & !enableGEM & !enableRPC ) backward = muonLayout->backwardCSCLayers(); //CSC only
   else backward = muonLayout->backwardCSCLayers();
```
Comment (slava77) [inline RecoMuon/Navigation/src/MuonNavigationPrinter.cc]: this looks wrong for the case of (enableCSC == true, enableGEM == false, enableRPC == true):
- the old code would give me allBackwardLayers, 
- the new code would give me backwardCSCLayers
I suspect this is the origin of the differences in current (run1/run2) reconstruction

**Comment by slava77** [comment]
this is a "-1", I would just like to keep it on our "pending" list

@archiesharma ,  sorry for the delay.

I'm looking at comparisons in jenkins (e.g. in ZMM MC in Run2 setup) in workflow 1330.0
https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_7_6_X_2015-07-08-2300+10109/6097/validateJR/all_OldVSNew_ZMM13TeVwf1330p0/
or in relmon
https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_7_6_X_2015-07-08-2300+10109/6097/1330.0_ZMM_13+ZMM_13+DIGIUP15+RECOUP15+HARVESTUP15+MINIAODMCUP15/Muons_RecoMuonV_RecoMuon_MuonAssoc_Sta.html
or 
https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_7_6_X_2015-07-08-2300+10109/6097/1330.0_ZMM_13+ZMM_13+DIGIUP15+RECOUP15+HARVESTUP15+MINIAODMCUP15/Muons_standAloneMuonsUpdatedAtVtx_GeneralProperties.html
)
there are differences that start apparently at the standalone muon reconstruction level

I suspect MuonNavigationSchool change, but it could be more than that.
Please double-check and submit an update.
Please test on some run1 or run2 data or MC with muons (I think even single muon events would be enough): you can just compare the standalone muon distributions by plotting variables interactively.

**Comment by archiesharma** [comment]
Pull request updated to solve the discrepancies in the standalone muons variables. The comparison plots were checked using Run 2 data and the they are coming same.

**Comment by slava77** [comment]
this PR topic branch was not updated appropriately.
rebase to a recent IB is needed to get rid of the merge commits in the history.

**Comment by archiesharma** [comment]
all comments are included in updated pull request (in CMSSW_7_6_X_2015-07-21-2300 ), please test/review

**Comment by slava77** [comment]
Hi Archie,
you can see a lot of irrelevant commits in this PR now https://github.com/cms-sw/cmssw/pull/10109/commits
This should be cleaned up by rebasing, or, alternatively by remaking the change in a new branch with a new commit (I don't think the history of changes is that essential).
We will never manage to get 10 signatures on this PR otherwise (only reco and dqm are real, the rest is the result of these irrelevant commits)

**Comment by ianna** [comment]
@archiesharma - please, rebase

**Comment by archiesharma** [comment]
Hi Slava,

The rebasing is done to update the branch and it is now contains all the latest commits. Please check  and test.

**Comment by slava77** [comment]
@archiesharma 
Nice. It looks good now.

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by deguio** [comment]
+1

**Comment by pietverwilligen** [comment]
The crashes in all the steps seems to be an HLT problem, not related to
this pull request ...

Step: HLT Spec: ['@relval25ns']

NameError: global name 'customiseFor10234' is not defined


Is this a known error for running the workflows. Do I understand well that
this is a customization for PR10234? Does this give an error because this
pull request is not integrated?
https://github.com/cms-sw/cmssw/pull/10234

This PR 10109 <https://github.com/cms-sw/cmssw/pull/10109> was tested
together with following PRs: 10311, 10268, 10239 (backport stage 1 unpacker
to 75X), 10310, 10303

The following merge commits were also included on top of IB + this PR after
> doing git cms-merge-topic:
> 0407562
> <https://github.com/cms-sw/cmssw/commit/0407562d6cf552c2ea38f0c55d5acde3e32a6312>
> 4b612a8
> <https://github.com/cms-sw/cmssw/commit/4b612a833ed77940f082eda8f31b6afea5398f41>
> 7cd869f
> <https://github.com/cms-sw/cmssw/commit/7cd869f325fd857ae248bf6d1ad9e68a09ea72d4>
> a201db4
> <https://github.com/cms-sw/cmssw/commit/a201db4d2fc9422415f23df11244b0b2559e9583>
> 550fdc9
> <https://github.com/cms-sw/cmssw/commit/550fdc9baf7c92d26c0bc1b8b2e76f8ba93a170c>
> b7655c6
> <https://github.com/cms-sw/cmssw/commit/b7655c60487938b8977110e29bfe6ae5ca784145>
> aca4d65
> <https://github.com/cms-sw/cmssw/commit/aca4d6546342ce6e8a7ce990857fdc36d278e40f>
>

Maybe 10234 (backport stage 1 unpacker to 74X) needs to be included as
well? Can some expert comment here?
Thanks
Piet


On Thu, Jul 23, 2015 at 10:39 PM, cmsbuild <notifications@github.com> wrote:

> -1
> Tested at: 0407562
> <https://github.com/cms-sw/cmssw/commit/0407562d6cf552c2ea38f0c55d5acde3e32a6312>
> When I ran the RelVals I found an error in the following worklfows:
> 4.53 step2
>
> runTheMatrix-results/4.53_RunPhoton2012B+RunPhoton2012B+HLTD+RECODreHLT+HARVESTDreHLT+MINIAODDreHLT/step2_RunPhoton2012B+RunPhoton2012B+HLTD+RECODreHLT+HARVESTDreHLT+MINIAODDreHLT.log
>
>
> 5.1 step1
>
> runTheMatrix-results/5.1_TTbar+TTbarFS+HARVESTFS/step1_TTbar+TTbarFS+HARVESTFS.log
>
>
> 135.4 step1
>
> runTheMatrix-results/135.4_ZEE_13+ZEEFS_13+HARVESTUP15FS+MINIAODMCUP15FS/step1_ZEE_13+ZEEFS_13+HARVESTUP15FS+MINIAODMCUP15FS.log
>
>
> 8.0 step2
>
> runTheMatrix-results/8.0_BeamHalo+BeamHalo+DIGICOS+RECOCOS+ALCABH+HARVESTCOS/step2_BeamHalo+BeamHalo+DIGICOS+RECOCOS+ALCABH+HARVESTCOS.log
>
>
> 1306.0 step2
>
> runTheMatrix-results/1306.0_SingleMuPt1_UP15+SingleMuPt1_UP15+DIGIUP15+RECOUP15+HARVESTUP15+MINIAODMCUP15/step2_SingleMuPt1_UP15+SingleMuPt1_UP15+DIGIUP15+RECOUP15+HARVESTUP15+MINIAODMCUP15.log
>
>
> 1330.0 step2
>
> runTheMatrix-results/1330.0_ZMM_13+ZMM_13+DIGIUP15+RECOUP15+HARVESTUP15+MINIAODMCUP15/step2_ZMM_13+ZMM_13+DIGIUP15+RECOUP15+HARVESTUP15+MINIAODMCUP15.log
>
>
> 9.0 step2
>
> runTheMatrix-results/9.0_Higgs200ChargedTaus+Higgs200ChargedTaus+DIGI+RECO+HARVEST/step2_Higgs200ChargedTaus+Higgs200ChargedTaus+DIGI+RECO+HARVEST.log
>
>
> 25.0 step2
>
> runTheMatrix-results/25.0_TTbar+TTbar+DIGI+RECOAlCaCalo+HARVEST+ALCATT/step2_TTbar+TTbar+DIGI+RECOAlCaCalo+HARVEST+ALCATT.log
>
>
> 1000.0 step3
>
> runTheMatrix-results/1000.0_RunMinBias2011A+RunMinBias2011A+TIER0+SKIMD+HARVESTDfst2+ALCASPLIT/step3_RunMinBias2011A+RunMinBias2011A+TIER0+SKIMD+HARVESTDfst2+ALCASPLIT.log
>
>
> 25202.0 step2
>
> runTheMatrix-results/25202.0_TTbar_13+TTbar_13+DIGIUP15_PU25+RECOUP15_PU25+HARVESTUP15_PU25+MINIAODMCUP15/step2_TTbar_13+TTbar_13+DIGIUP15_PU25+RECOUP15_PU25+HARVESTUP15_PU25+MINIAODMCUP15.log
>
>
> 50202.0 step2
>
> runTheMatrix-results/50202.0_TTbar_13+TTbar_13+DIGIUP15_PU50+RECOUP15_PU50+HARVESTUP15_PU50+MINIAODMCUP1550/step2_TTbar_13+TTbar_13+DIGIUP15_PU50+RECOUP15_PU50+HARVESTUP15_PU50+MINIAODMCUP1550.log
>
>
> you can see the results of the tests here:
>
> https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-10109/6440/summary.html
>
> The following merge commits were also included on top of IB + this PR
> after doing git cms-merge-topic:
> 0407562
> <https://github.com/cms-sw/cmssw/commit/0407562d6cf552c2ea38f0c55d5acde3e32a6312>
> 4b612a8
> <https://github.com/cms-sw/cmssw/commit/4b612a833ed77940f082eda8f31b6afea5398f41>
> 7cd869f
> <https://github.com/cms-sw/cmssw/commit/7cd869f325fd857ae248bf6d1ad9e68a09ea72d4>
> a201db4
> <https://github.com/cms-sw/cmssw/commit/a201db4d2fc9422415f23df11244b0b2559e9583>
> 550fdc9
> <https://github.com/cms-sw/cmssw/commit/550fdc9baf7c92d26c0bc1b8b2e76f8ba93a170c>
> b7655c6
> <https://github.com/cms-sw/cmssw/commit/b7655c60487938b8977110e29bfe6ae5ca784145>
> aca4d65
> <https://github.com/cms-sw/cmssw/commit/aca4d6546342ce6e8a7ce990857fdc36d278e40f>
> You can see more details here:
>
> https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-10109/6440/git-log-recent-commits
>
> https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-10109/6440/git-merge-result
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/10109#issuecomment-124233878>.
>



-- 
Piet Verwilligen

INFN -- Sezione di Bari
Via E. Orabona 4
I-70125 Bari, Italy
Phone: +39 345 74 70 642

**Comment by slava77** [comment]
the IB used for tests is indeed broken (CMSSW_7_6_X_2015-07-23-2300). Hopefully the next one is good.

**Comment by slava77** [comment]
@cmsbuild please test
the latest IB (at least CMSSW_7_6_X_2015-07-24-2300 had the matrix tests running OK)

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/10109	0407562
- the forward port is done appropriately (no interference with the current run1/run2 reco/simulation) as confirmed in jenkins test
- jenkins tests pass and comparisons with the baseline show no differences

**Comment by davidlange6** [comment]
+1

**Pull request merged**